controller/aks-cluster-config-handler.go

```diff
@@ -416,6 +416,24 @@ func (h *Handler) validateConfig(config *aksv1.AKSClusterConfig) error {
 	if to.String(config.Spec.NetworkPolicy) == string(containerservice.NetworkPolicyAzure) && to.String(config.Spec.NetworkPlugin) != string(containerservice.Azure) {
 		return fmt.Errorf("azure network policy can be used only with Azure CNI network plugin for [%s] cluster", config.ClusterName)
 	}
+	cannotBeNilErrorAzurePlugin := "field [%s] must be provided for cluster [%s] config when Azure CNI network plugin is used"
+	if to.String(config.Spec.NetworkPlugin) == string(containerservice.Azure) {
+		if config.Spec.VirtualNetwork == nil {
+			return fmt.Errorf(cannotBeNilErrorAzurePlugin, "virtualNetwork", config.ClusterName)
+		}
+		if config.Spec.Subnet == nil {
+			return fmt.Errorf(cannotBeNilErrorAzurePlugin, "subnet", config.ClusterName)
+		}
+		if config.Spec.NetworkDNSServiceIP == nil {
+			return fmt.Errorf(cannotBeNilErrorAzurePlugin, "dnsServiceIp", config.ClusterName)
+		}
+		if config.Spec.NetworkDockerBridgeCIDR == nil {
+			return fmt.Errorf(cannotBeNilErrorAzurePlugin, "dockerBridgeCidr", config.ClusterName)
+		}
+		if config.Spec.NetworkServiceCIDR == nil {
+			return fmt.Errorf(cannotBeNilErrorAzurePlugin, "serviceCidr", config.ClusterName)
+		}
```

Comment:
What if a cluster was registered and then the NetworkPlugin was changed? Would these errors be encountered?

Reply:
I think it's safe because this will bail early if it's a registered cluster https://github.com/rancher/aks-operator/pull/33/files#diff-1aa54d4941ca80708078867e1b130aff086719689479dea38683fded2f3c7852R362-R364

But for GKE and EKS we require all top-level parameters are set, even if they're empty and not going to be used. Should we be doing that here as well, regardless of the network plugin?

Reply:
After cluster was provisioned, NetworkPlugin can't be changed. This applies to the most hosted Kubernetes clusters.